source/client/service_main.cc

```diff
@@ -49,13 +67,53 @@ void ServiceMain::start() {
     absl::StrAppend(&listener_bound_address_, listener_port_);
   }
   ENVOY_LOG(info, "Nighthawk grpc service listening on: {}", listener_bound_address_);
+  if (listener_output_path_ != "") {
+    std::ofstream myfile(listener_output_path_);
+    if (myfile.is_open()) {
+      myfile << listener_bound_address_;
+      myfile.close();
+    }
+  }
   channel_ = grpc::CreateChannel(listener_bound_address_, grpc::InsecureChannelCredentials());
   stub_ = std::make_unique<nighthawk::client::NighthawkService::Stub>(channel_);
+  pipe_fds_.resize(2);
+  // The shutdown thread will be notified of by our signal handler and take it from there.
+  RELEASE_ASSERT(pipe(pipe_fds_.data()) == 0, "pipe failed");
+
+  shutdown_thread_ = std::thread([this]() {
+    int tmp;
+    RELEASE_ASSERT(read(pipe_fds_[0], &tmp, sizeof(int)) >= 0, "read failed");
+    RELEASE_ASSERT(close(pipe_fds_[0]) == 0, "read side close failed");
+    RELEASE_ASSERT(close(pipe_fds_[1]) == 0, "write side close failed");
+    pipe_fds_.clear();
+    server_->Shutdown();
+  });
 }
 
-void ServiceMain::wait() { server_->Wait(); }
+void ServiceMain::wait() {
+  signal_handler_delegate = [this](int) { onSignal(); };
+  signal(SIGTERM, signal_handler);
```

Comment:
Do we want to mix raw signals and libevent like this? I thought libevent has dedicated APIs for signal handling.

Reply:
Right; looking into that I think ideally we'd use `DispatcherImpl::listenForSignal()`..

But we'd need a `DispatcherImpl` instance - or a means to allocate one - to leverage that, and I can't see a convenient way to accomplish either. This file just wraps/manages the Nighthawk gRPC service, and has no ties to NH/Envoy concepts (yet) otherwise.. 
If I'm correct, pulling this of would require wiring a way to request signal handling up accross the gRPC service implementation in `ServiceImpl` and then `ProcessImpl` where the main dispatcher lives.

when I scanned Envoy's codebase, I couldn't find any other signal hooking going on except in Envoy's server.cc - which will not run concurrently to this, so based on that this seemed viable to me (as an intermediate step at least).

Reply:
Ah, right, that's fair.